datacube/virtual/impl.py

```diff
@@ -280,16 +294,18 @@ def query(self, dc, **search_terms):
             datasets = select_datasets_inside_polygon(dc.index.datasets.search(**query.search_terms),
                                                       query.geopolygon)
 
+            # should we put it in the Transformation class?
             if get('dataset_predicate') is not None:
                 datasets = [dataset
                             for dataset in datasets
                             if get('dataset_predicate')(dataset)]
 
             # gather information from the index before it disappears from sight
             # this can also possibly extracted from the product definitions but this is easier
-            grid_spec = dc.index.products.get_by_name(self._product).grid_spec
+            product = dc.index.products.get_by_name(self._product)
```

Comment:
pretty sure `product` could be `None`, should probably `raise ValueError` if it is

Reply:
Can't be `None` here because if the product is not present, then `dc.index.datasets.search` fails with the message `ValueError: No products match search terms: {'product': 'non-existant'}`, which is pretty much what you are asking for.

Reply:
Too many indirections to be certain, do an `assert` then if validation of the product is done elsewhere, or better grab an actual product object at the site that does validation

Reply:
I would just re-order this, get product first, if None raise error to user that makes sense -- they supplied wrong product name, and only if successful do query